File: Solutions/677. Map Sum Pairs/677. Map Sum Pairs.java
public class trie{
    trie[] ch=new trie[26];
    int sum=0;
}
class MapSum {
    private Map<String,Integer> m=new HashMap<>();
    private trie root=new trie(); //here  ,java always builds object to become a reference
    /** Initialize your data structure here. */
    public MapSum() {
        
    }
    
    public void insert(String key, int val) {
        trie p=root; //here, var like auto ,or "trie p=root" is OK
        for(var c:key.toCharArray()){
            if(p.ch[c-'a']!=null){
                p.ch[c-'a']=p.ch[c-'a'];
                p=p.ch[c-'a'];
            }else{
                p.ch[c-'a']=new trie();
                p=p.ch[c-'a'];
            }
            p.sum+=val-m.getOrDefault(key,0);
        }
        m.put(key,val);
    }
    
    public int sum(String prefix) {
        trie p=root;
        for(var c:prefix.toCharArray()){
            if(p==null)
                break;
            p=p.ch[c-'a'];
        }
        return p==null?0:p.sum;
    }
}

/**
 * Your MapSum object will be instantiated and called as such:
 * MapSum obj = new MapSum();
 * obj.insert(key,val);
 * int param_2 = obj.sum(prefix);
 */

File: Solutions/677. Map Sum Pairs/677. Map Sum Pairs.py
class Trie:
    def __init__(self):
        self.ch=defaultdict(Trie)
        self.sum=0
class MapSum:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root=Trie()
        self.m=defaultdict(int)

    def insert(self, key: str, val: int) -> None:
        p=self.root
        for c in key:
            if p.ch[c]!=None:
                p.ch[c]=p.ch[c]
                p=p.ch[c]
            else:
                p.ch[c]=Trie()
                p=p.ch[c]
            p.sum+=val-self.m[key]
        self.m[key]=val

    def sum(self, prefix: str) -> int:
        p=self.root
        for c in prefix:
            if p==None:
                break;
            p=p.ch[c]
        return p.sum if p!=None else 0

# Your MapSum object will be instantiated and called as such:
# obj = MapSum()
# obj.insert(key,val)
# param_2 = obj.sum(prefix)